sycl/doc/CompileTimeProperties.md

```diff
@@ -0,0 +1,813 @@
+* `[[__sycl_detail__::add_ir_function_attributes()]]`
+
+When the front-end consumes these C++ attributes and produces IR, each property
+name becomes an IR attribute name and each property value becomes the
+attribute's value.  Because the attribute values must be strings, the front-end
+converts each property value to a string.  Integer and floating point values
+are converted with the same format as `std::to_string()` would produce.
+Boolean values are converted to either `"true"` or `"false"`.  Enumeration
+values are first converted to an integer and then converted to a string with
+the same format as `std::to_string()`.  The `nullptr` value is converted to an
+empty string (`""`).
+
+**TODO**: Should we allow property values that are type names?  If so, I
+suppose they would be converted to a string representation of the mangled name?
+
+**TODO**: Should we allow property values of other (non-fundamental) types?  If
+we allow this, we need to teach the front-end how to convert each type to a
+string, which means the front-end needs to be changed each time we add a
+property with a new non-fundamental type.  This seems undesirable.  However, if
+we do not allow non-fundamental types, how do we represent properties like
+`work_group_size`, whose value is a 3-tuple of integers?  Maybe we could just
+allow `std::tuple`, where the type of each element is one of the fundamental
+types listed above.
+
+### IR representation via `@llvm.ptr.annotation`
+
+Properties that are implemented using
+`[[__sycl_detail__::add_ir_member_annotation()]]`, are represented in LLVM IR
+as the fifth metadata parameter to the `@llvm.ptr.annotation` intrinsic
+function.  This parameter is a tuple of metadata values with the following
+sequence:
+
+* Name of the first property
+* Value of the first property
+* Name of the second property
+* Value of the second property
+* Etc.
+
+Since metadata types are not limited to strings, there is no need to convert
+the property values to strings.
+
+
+## Filtering properties
+
+It is sometimes necessary to filter out certain properties so that only a
+subset of the properties in a list are represented in IR.  There are two
+scenarios when this is useful.
+
+In some cases, a property is used only in the header file itself, and there is
+no need to represent that property in LLVM IR.  In order to avoid cluttering
+the IR with unneeded information, these properties can be "filtered out", so
+that the front-end does not generate an IR representation.
+
+Another case is when a class wants to represent some properties one way in the
+IR while representing other properties in another way.  For example, a future
+version of `accessor` might pass some properties to
+`[[__sycl_detail__::add_ir_kernel_parameter_attributes()]]` while passing other
+properties to `[[__sycl_detail__::add_ir_member_annotation()]]`.  Again, the
+header wants some way to "filter" the properties, such that some properties are
```

Comment:
Sorry I don't understand this use-case. What does the pass list have to do with this? Won't the fact that 2 different attributes are being used indicate whether the property should be interpreted as kernel parameter properties vs member annotations

Reply:
The example further down in this section illustrates this use case.  See the example after "To illustrate, consider the following example ...".  Does that answer your question?